### DISCUSSION TITLE
Access to path denied by write to program files 'appsettings.json' by custom action (Wix4)

### DISCUSSION THREAD

**Discussion by marrale** [wixtoolset/issues#7990] [Questions] [answered]
My Wix4 msi installer, at the end of the installation process, updates "Program Files/MyApp/appsettings.json" values, from filled out fields, in the Installer UI dialog.

If I execute the installer by command prompt run as "administrator" everything works as expected.
But if I execute my msi file normally once the "appsettings.json" following error is thrown:
"Access to the path 'C:\Program Files\MyApp\appsettings.json' is denied."

I'm searching for hours now, but unsuccessfully.

Found suggestions:
- Add `Impersonate="no"` attribute to "CustomAction" element -> done
- Add `Scope="perMachine"` attribute to "Package" element -> done
- Add `<Property Id="MSIUSEREALADMINDETECTION" Value="1"  />` Element to package Element -> done
- Add `InstallPrivileges="elevated"` attribute to "Package" element -> this attribute not exists in Wix4 Package element

Any other hint how to get the custom action executed with elevated privileges to get my installer works correctly?
If I create an Installation Bundle that executes my msi everything works as expected. Is this the only way?

PS: the user that executes the installer has admin rights

My Package:
```
<Package Name="$(MyProductNameProperty)" Manufacturer="MyCompany" Version="1.0.0.0" UpgradeCode="3cada225-3b4e-4612-9dca-297b8155c6a2" Compressed="yes" Scope="perMachine" ProductCode="*" >
    <Property Id="MSIUSEREALADMINDETECTION" Value="1"  />
    ...
</Package>
```

My custom action reference:
```
<CustomAction Id='UpdateAppSettings' BinaryRef='AppSettingsCA' DllEntry='UpdateAppSettings' Impersonate="no" />
<Binary Id='AppSettingsCA' SourceFile='..\AppSettingsCustomAction\bin\Release\net472\AppSettingsCustomAction.CA.dll' />
```

Installation package sequence:
```
<InstallExecuteSequence>
    <!-- Need to be executed after the files was copied https://wixtoolset.org/docs/tools/wixext/util/ -->
    <Custom Action="UpdateAppSettings" After="InstallFinalize" Condition="NOT Installed" />
</InstallExecuteSequence>
```

Part of the custom action:
```
[CustomAction]
public static ActionResult UpdateAppSettings(Session session)
{
    session.Log("Begin UpdateAppSettings");
    string path = session["INSTALLFOLDER"];
    string filePath = $@"{path}appsettings.json";
    session.Log($"File Path: '{filePath}'");
    string node = session["URL"];

    ...

    File.WriteAllText(filePath, output);
    return ActionResult.Success;
}
```

Instead to update an xml file, it can be very simply be done by wix utils:
```
<ComponentGroup Id="UpdateConfigComponents" Directory="INSTALLFOLDER">
    <Component Id="UpdateConfigComponent" Guid="c66c0b04-aab5-474f-8fbf-7aae6e40b0f4">
        <!-- https://wixtoolset.org/docs/v3/xsd/util/xmlfile/ -->
        <util:XmlFile
            Id="UpdateUrl"
            Action="setValue"
            File="[INSTALLFOLDER]MyApp.dll.config"
            SelectionLanguage="XPath"
            Permanent="yes"
            ElementPath="/configuration/appSettings/add[\[]@key='Url'[\]]"
            Name="value"
            Value="[URL]" />
    </Component>
</ComponentGroup>
```

Eventually it exists a similar solution to update my "appsettings.json"? Or I should my webapi read the url from an xml file instead of the json file?

Thanks.

**Comment by nirbar** [comment, marked as answer]
The custom action must have Execute="deferred" and scheduled after InstallInitialize and before InstallFinalize.
Using util XmlFile or my extension (see below) also makes sure that the file is properly rolled back on errors

To use my WiX extension https://www.nuget.org/packages/PanelSwWixExtension4:
```
      <Binary SourceFile="$(sys.SOURCEFILEDIR)XSLTFile1.xslt" Id="XslTransform" />

      <Component>
        <File Source="$(sys.SOURCEFILEPATH)">
          <PanelSW:XslTransform BinaryKey="XslTransform">
            <PanelSW:Replace Text="\[\[]FILE_NAME\[\]]" Replacement="[FILE_NAME]" />
          </PanelSW:XslTransform>
        </File>
      </Component>
```

**Reply by marrale** [reply]
@nirbar thanks for your answer, setting up the custom action as Execute="deferred" as the trick. Now it works as expected.
